src/wallet/wallet_rpc_server.cpp

```diff
@@ -3089,6 +3089,211 @@ namespace tools
+    std::unique_ptr<wallet2> wal;
+    wal = std::move(rc.first);
+    if (!wal)
+    {
+      er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+      er.message = "Failed to create wallet";
+      return false;
+    }
+
+    epee::wipeable_string password = rc.second.password();
+   
+
+    bool was_deprecated_wallet = ((old_language == crypto::ElectrumWords::old_language_name) ||
+    crypto::ElectrumWords::get_is_old_style_seed(req.seed));
+
+    std::string mnemonic_language = old_language;
+    std::string m_mnemonic_language;
+    std::vector<std::string> language_list;
+    crypto::ElectrumWords::get_language_list(language_list);
+    if (mnemonic_language.empty() && std::find(language_list.begin(), language_list.end(), m_mnemonic_language) != language_list.end())
+    {
+      mnemonic_language = m_mnemonic_language;
+    }
+
+    if (((mnemonic_language.empty() || mnemonic_language == crypto::ElectrumWords::old_language_name) && was_deprecated_wallet))
+    {
+      if (was_deprecated_wallet)
+      {
+        // The user had used an older version of the wallet with old style mnemonics.
+        res.was_deprecated = true;
+        return false;
+      }
+      if (req.language.empty())
+      {
+        er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+        er.message = "Wallet is deprecated. Would provide new seed. Please specify language to use for new seed.";
+        return false;
+      }
+      mnemonic_language = req.language;
+    }
+
+    wal->set_seed_language(mnemonic_language);
+
+    crypto::secret_key recovery_val;
+    try
+    {
+      recovery_val = wal->generate(m_wallet_file, std::move(rc.second).password(), m_recovery_key, true, false, false);
+      MINFO("Wallet has been restored.\n");
+    }
+    catch (const std::exception& e)
+    {
+      handle_rpc_exception(std::current_exception(), er, WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR);
+      return false;
+    }
+
+    // // Convert the secret key back to seed
+    epee::wipeable_string electrum_words;
+    if (!crypto::ElectrumWords::bytes_to_words(recovery_val, electrum_words, mnemonic_language))
+    {
+      er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+      er.message = "Failed to encode seed";
+      return false;
+    }
+    res.seed = electrum_words.data();
+
+    if (!wal)
+    {
+      er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+      er.message = "Failed to generate wallet";
+      return false;
+    }
+
+    // set blockheight if given
+    try
+    {
+      if (req.restore_height)
+      {
+        wal->set_refresh_from_block_height(req.restore_height);
+        wal->rewrite(m_wallet_file, password);
+      }
+      else
+      {
+        cryptonote::COMMAND_RPC_GET_HEIGHT::request hreq;
+        cryptonote::COMMAND_RPC_GET_HEIGHT::response hres;
+        hres.height = 0;
+        bool r = wal->invoke_http_json("/getheight", hreq, hres);
+        wal->set_refresh_from_block_height(hres.height);
```

Comment:
I don't see the point of this behavior; when restoring a wallet from seed, most likely the user has already made several transactions with the wallet previously (hence "restoring"), so setting the restore height to the current blockchain height seems useless. The CLI wallet defaults the restore height to zero if the user didn't specify otherwise.